CMakeLists.txt

```diff
@@ -125,27 +125,32 @@ endif()
 #----------------------------------------------------------
 # CUDA support
 #----------------------------------------------------------
-option(ARB_WITH_CUDA "use CUDA for GPU offload" OFF)
-if(ARB_WITH_CUDA)
-    find_package(CUDA REQUIRED)
-
-    # Turn off annoying and incorrect warnings generated in the JSON file.
-    # We also work around the same issue with the intel compiler.
-    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-Xcudafe \"--diag_suppress=not_used_in_template_function_params\";-Xcudafe \"--diag_suppress=cast_to_qualified_type\")
+set(ARB_GPU_TARGET "none" CACHE STRING "The target GPU architecture: one of {none,K20,K80,P100}")
+set_property(CACHE ARB_GPU_TARGET PROPERTY STRINGS none K20 K80 P100 )
 
-    # set the CUDA target specfic flags
-    # code regions protected by ARB_HAVE_CUDA should only be available to the CUDA
-    # compiler, which regions protected by ARB_HAVE_GPU are visible to both host
-    # and device compiler when targetting GPU.
-    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-DARB_HAVE_CUDA)
-    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-DARB_HAVE_GPU)
-    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-arch=sm_60) # minimum target P100 GPUs
+set(ARB_WITH_CUDA FALSE)
 
+if(NOT ARB_GPU_TARGET MATCHES "none")
+    find_package(CUDA REQUIRED)
+    set(ARB_WITH_CUDA TRUE)
     add_definitions(-DARB_HAVE_GPU)
     include_directories(SYSTEM ${CUDA_INCLUDE_DIRS})
     list(APPEND EXTERNAL_LIBRARIES ${CUDA_LIBRARIES})
 endif()
 
+if(ARB_GPU_TARGET MATCHES "K20")
+    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-arch=sm_35)
+    add_definitions(-DARB_CUDA_ARCH=350)
```

Comment:
The `ARB_CUDA_ARCH` approach of determining if synchronization is required or not is a bit delicate: in principle, the CUDA code could be compiled to support multiple GPU architectures, and then the calling code would have to pick the lowest common denominator for the synchronization test, or it could be a property of the GPU execution context when that proposed work is integrated.

Could we set this configuration option up as describing the minimal virtual architecture to generate code for — for the purposes of the synchronization test — and then have a separate option for configuring the actual architectures used for code generation and assembly, which could default to either (say) the supplied minimal architecture, or encompass K20, K80 and P100 targets?

Reply:
The current approach for CUDA is "target the following GPU model (or target GPU architecture if you get your way with the above suggestion)", and this approach reflects that, even if it isn't the nicest design.

The (very) near future plan is to have a "GPU Context" that would pick at runtime pick kernels etc. compiled for the GPU that is provided to the library. Can we defer the decoupling until then, when it will be quite natural?

Reply:
Sure, fair call. We can debate subtleties then.